### ISSUE TITLE
[Feature Request] inferred module dimensions

### ISSUE THREAD

**Issue by bethebunny** [pytorch/pytorch#23352]
## Motivation

There’s a lot of cases in framework design where we want to be able to treat components like lego blocks when building various model architectures. The following case studies in PyText are the 90% common cases, not the exceptions, ie. most people build models from these architecture families.

Document classifiers in PyText follow a general pattern of Embedding → Representation → Decoder.

* Embedding
    * Word embedding
    * BiTransformer
    * Character embedding
* Representation
    * Encoder
        * DocNN
        * LSTM (possibly bidirectional)
    * Pooling layer
        * Max pool
        * Attention
        * etc
* Decoder
    * Almost always an MLP, needs to have an output dimension of #classes

Within the 4 categories of Embedding/Encoder/Pooling/Decoder though, each boundary depends on the dimensionality of the previous component — the encoder depends on the embedding dimension, the pooling depends on the encoder dimension (including eg. whether the LSTM is bidirectional), and the decoder depends on the pooling layer output dimension (and #classes).

However, from a user standpoint, it is extremely desirable to not have to make sure that these dimensions match up exactly. For instance, if a product team wants to experiment with DocNN vs BiLSTM with self attention and two different sizes of word embeddings, they would need to know how to compute the correct input dimensions for their encoders, pooling layers, and MLPs for each of the 4 cases. If they have dense features, those might also factor into the computation for the MLP layer.

PyText solves this problem by having modules of each of these categories compute and track the appropriate dimensionality internally to their constructor, and then constructing the model and wiring the various blocks together before training. See [DocNN](https://github.com/facebookresearch/pytext/blob/a64370c7f1aac1c0491ccf5b2b0987909c9e0d94/pytext/models/representations/docnn.py#L30) representation_dim computation for an example of this, and [Model.from_config](https://github.com/facebookresearch/pytext/blob/a64370c7f1aac1c0491ccf5b2b0987909c9e0d94/pytext/models/model.py#L167) for where we use this to construct the model.

This solution works for us, but only in the case where we take the responsibility of constructing the model, along with all of its subcomponents, away from the user. If we wanted to eg. allow the user to pass in their own representation layer, we wouldn’t be able to ensure the dimensionality match, the user is therefore responsible for also constructing _every other model component_. As a direct result of this design, PyText is abysmally hard to use in a notebook, and we built a Config system for managing hyperparameters that essentially boils down to allowing users to partially construct modules, then let us take over and finish building the whole model. See one of our [BiLSTM Config](https://github.com/facebookresearch/pytext/blob/master/pytext/models/representations/bilstm_doc_attention.py#L35) classes for an example.

This issue is not specific to PyText. Consider this snippet from the demo example from the AllenNLP website:

```python
EMBEDDING_DIM = 6
HIDDEN_DIM = 6

token_embedding = Embedding(num_embeddings=vocab.get_vocab_size('tokens'),
embedding_dim=EMBEDDING_DIM)
word_embeddings = BasicTextFieldEmbedder({"tokens": token_embedding})
lstm = PytorchSeq2SeqWrapper(torch.nn.LSTM(EMBEDDING_DIM, HIDDEN_DIM, batch_first=True))
model = LstmTagger(word_embeddings, lstm, vocab)
```
(source - https://allennlp.org/tutorials)

Note that similar to PyText, we need to match up the EMBEDDING_DIM between token_embedding and the LSTM module; we therefore couldn’t construct an LstmTagger object with a default argument for just one of word_embeddings or lstm, because there would be no way to create them with the appropriate dimensions.

## Proposal - Lazy Modules

We propose adding lazily inferred dimensions to some core Torch components, such as Linear, ConvNd, and RNN. This syntax should allow constructing these components (and any derived or more complex user components containing them) to instantiate the models without fully defining their dimensions, and have them inferred from the tensors passed in the first time their forward function is called. We propose using -1 to represent these dimensions similar to how -1 is used in reshape now (None has also been suggested, covered in alternatives below). For example

```python
>>> l = nn.Linear(-1, 3)
>>> l(torch.rand(2, 4)).size()
[2, 3]
>>> l(torch.rand(2, 5))
TypeError(...)
```

This would allow building more complex components that could also infer their input dimensions; consider a simple MLP implementation, much like what PyText uses for its decoders:

```python
class MLP(nn.Module):
    def __init__(self, dims: List[int], dropout: float = 0.):
        super().__init__()
        layers = []
        for in_dim, dim in zip([-1] + dims, dims):
            layers.extend([
                nn.Linear(in_dim, dim),
                ReLU(),
                nn.LayerNorm(dim),
                nn.Dropout(dropout),
            ])
        self.layers = nn.Sequential(*layers)

    def forward(self, input: torch.Tensor) -> torch.Tensor:
        return self.layers(input)
```

Now this MLP can be used in any PyText model, regardless of what the output shape of representation/pooling is:

```python
mlp1 = MLP([4, 5, 6])
mlp2 = MLP([4, 5, 6])

mlp1(torch.rand(2, 4))  # finalizes mlp1 as expecting input dimension 4
mlp2(torch.rand(2, 5))  # finalizes mlp2 as expecting input dimension 5
```

And in the cases of PyText and AllenNLP these could allow interfaces that are much simpler and more modular:

```python
class Model(nn.Module):
    def __init__(
        self,
        embedding = WordEmbedding(),
        representation = BiLSTM(),
        decoder = MLP([4, 5, 6]),
    ):
        self.embedding = embedding
        self.representation = representation
        self.decoder = decoder
    
    def forward(self, tokens):
        embedded = self.embedding(tokens)
        representation = self.representation(embedded)
        return self.decoder(representation)
        
model1 = Model()
model2 = Model(representation=DocNN())
model3 = Model(
    embedding=WordEmbedding(embedding_dim=200, vocab_size=10000),
    decoder=MLP([2, 50, 10]),
)
```

## Sample implementation

This will be supplied shortly following this proposal as a pull request in torch.nn.lazy, but for now here is the proposed implementation.

```python
from torch import nn

# LazyModuleMeta re-implements the type construction semantics for objects to allow
# a slight variant on syntax. Essentially anything with this metaclass can optionally
# execute a single yield statement during its constructor (normal constructors also work fine).
# If it does yield during construction, then a __lazy_init__ function is populated;
# any code occurring before yield in the constuctor will be called as normal during object creation,
# and any code after yield will instead be deferred to the first call of __lazy_init__.

class LazyInitMeta(type):
    def __call__(cls, *args, **kwargs):
        if hasattr(cls, '__new__'):
            obj = cls.__new__(cls, *args, **kwargs)
        else:
            obj = object.__new__(cls)
        
        def initialize(obj):
            res = obj.__init__(*args, **kwargs)
            if isinstance(res, types.GeneratorType):
                next(res, None)
                def lazy_init(call_args):
                    try:
                        res.send(call_args)
                    except StopIteration:
                        pass
                    finally:
                        obj.__lazy_init__ = None
                obj.__lazy_init__ = lazy_init
            else:
                obj.__lazy_init__ = None
            
        if isinstance(obj, cls):
            initialize(obj)
        return obj
        
# Here's a Lazy nn.Module implementation using LazyInitMeta, calling __lazy_init__ before the first
# forward pass.
    
class LazyModule(nn.Module, metaclass=LazyInitMeta):
    def __init__(self):
        nn.Module.__init__(self)
    
    def __call__(self, *args, **kwargs):
        if self.__lazy_init__:
            self.__lazy_init__(call_args=(args, kwargs))
        return nn.Module.__call__(self, *args, **kwargs)
        
# Optionally lazy Linear module, based on the current torch implementation of nn.Linear.

class Linear(nn.Linear, LazyModule):
    def __init__(self, in_features, out_features, bias=True):
        LazyModule.__init__(self)
        self.in_features = in_features
        self.out_features = out_features
        if bias:
            self.bias = nn.Parameter(torch.Tensor(out_features))
        else:
            self.register_parameter('bias', None)
            
        if in_features == -1:
            self.register_parameter('weight', None)
            ([input], _) = yield  # lazy init remainder
            in_features = self.in_features = input.size()[-1]
            
        self.weight = nn.Parameter(torch.Tensor(out_features, in_features))
        self.reset_parameters()
```

## Drawbacks

Lazy modules don’t work perfectly with everything in torch. In particular, if they’re not yet finalized (ie. -1 was passed to a dimension to be inferred, but forward hasn’t been called yet), the following behaviors might be unintuitive:

* torch.save/torch.jit.script won’t work properly
    * both confirmed to work fine with the sample implementation once initialization finishes
* module.apply()
* any parameter initialization functions
* module.parameters() won’t return all of the model parameters, so
    * creating an optimizer/scheduler normally, ie torch.optim.Adam(module.parameters()) will potentially have unintended side-effects, as not all of the module’s eventual parameters will optimized
    * calling module.to() will move the parameters that exist, but once forward is called and new parameters are created, they will be on the default device
        * this can potentially be solved with some tighter init behavior and unit tests, as there’s a finite number of components that need to be made lazy, and they can usually look at other parameters on the module and assume that users want them to be on the same device
        * alternatively it may make sense to have a concept of module-level device residency

### None instead of -1

Use the above proposal, but pass None to indicate an inferred dimension rather than -1. -1 was chosen for the proposal because it mirrors the semantics of reshape and a few other operators, but None may be more intuitive.

### Support -1 (or None) for Tensor dimensions

Instead of implementing this at the Module level with lazy modules, implement this at the Tensor level. Have Tensors be able to be lazy- or partially-initialized, and then be able to be finalized with their final dimensions. This would solve most of the above drawbacks above, as for instance we’d always be making and setting all parameters, and it would also likely simplify its integration as all known useful cases of lazy dimensions eventually boil down to passing them to Tensors. However, it’s less clear what this implementation would look like, for instance how the Tensors would generically be able to infer the parameters.

### Lazy Parameters

If this were implemented at the Parameter level, it would require pretty radically increasing the complexity of how Parameters work, but could solve most of the drawbacks of this proposal while also leaving Tensor untouched. We can flesh out this design more if there is significant feedback in that direction.

### Support via metaclass vs non-metaclass

The current proposal implementation uses a metaclass for Lazy modules to introduce a novel initialization pattern for lazy components (this same pattern could eg. also be used to solve an analogous problems for Optimizer/Scheduler in the future). However, using metaclasses always comes at a cost in Python, especially for libraries that are being built on-top-of, because multiple inheritance in the case of multiple metaclasses is a problem with no generic solution. In other words, if any libraries built on top of pytorch want to use metaclasses in their own Module subclasses, they’ll need to do non-trivial work to make sure those metaclasses are compatible with any metaclass used in pytorch.
The proposal could likely be modified to, for example, use a decorator on the __init__ function instead of a metaclass.

**Comment by soumith** [comment]
This has come up several times in the past, and I spent the last 20 mins searching for the proposal I wrote on a previous issue, but I wasn't able to (the fun of having thousands of issues).
Hence, I'm re-iterating my previous thoughts both from memory and after re-thinking.

I think this is a feature worth designing, but the following design principles have to be upheld:
- no meta-programming or added magic to support it
- the module hard-errors on all cases where it is not fully initialized, with a clear message stating to the user what is expected of them
- the Auto-Infer mechanism has to be explicitly requested by the user, not via constants such as `-1` or `None`, but something like `torch.nn.Linear(torch.nn.INFER, 32)`.

To flesh out the implementation details,
- something like `nn.Linear`, when it gets as constructor input a `torch.nn.INFER`, will explicitly set it's weight to `nn._UninitializedParameter()`.
- the downstream tasks that handle parameters have to now explicitly handle and throw exceptions appropriately on getting `nn._UninitializedParameter`, with a clear error message
  - for example `.parameters()` should throw, `.apply()` should throw, `jit` entrypoints should throw, `.to()` doesn't need to throw.
- when a user does `model(input)`, `nn.Linear` will flesh out it's uninitalized parameters
- the performance hit of this deferred initialization logic should not affect the non-deferred initialization at all

**Comment by bethebunny** [comment]
Ok, I'm playing around with some implementations of UninitializedParameter that seem promising. One approach that I want to call out early because I don't necessarily understand all of the implications is to make use of reshape_(). If we did this, we could actually allow parameters() for instance to return correct references to the tensors, and things like .to() and .apply() should also just work. Are there other issues with this approach that I haven't run into yet?

Another issue is generally parameter initialization, eg. init.kaiming_uniform_, etc. I think the clever way to do this would be to somehow track a stack of initialization ops until the parameter is finalized and execute them then, and the easy and possibly just better solution is to somehow have these ops instead fail if the parameter hasn't been finalized (although this will definitely require downstream changes to how people initialize parameters in some cases).

**Comment by bethebunny** [comment]
The basic sketch of this implementation would look like

```
import torch
from torch import nn
from torch.nn import Parameter

INFER = object()

class UninitializedParameter(Parameter):
    def __new__(cls, infer_size_fn, dtype=torch.float, requires_grad=True):
        dummy_data = torch.tensor([], dtype=dtype)
        return nn.Parameter.__new__(cls, dummy_data, requires_grad)
    
    def __init__(self, infer_size_fn, dtype=torch.long):
        self.infer_size_fn = infer_size_fn
        
    def finalize(self, *args, **kwargs):
        requires_grad, self.requires_grad = self.requires_grad, False
        self.resize_(*self.infer_size_fn(*args, **kwargs))
        delattr(self, "infer_size_fn")
        self.requires_grad = requires_grad
        
    # size?
    # uniform_, et al?
    # __getitem__/__setitem__?

# This logic could just be incorporated into nn.Module, but it's pulled out here
# for easy notebook development

class ModuleWithUninitialized(nn.Module):
    def __init__(self):
        super().__init__()
        self._finalize_handle = self.register_forward_pre_hook(self._finalize)
        
    def _finalize(self, *input):
        for key, parameter in self._parameters.items():
            if isinstance(parameter, UninitializedParameter):
                parameter.finalize(*input)
                self._parameters[key] = Parameter(parameter.data, parameter.requires_grad)
        if hasattr(self, "reset_parameters"):
            self.reset_parameters()
        self._finalize_handle.remove()
        delattr(self, "_finalize_handle")
        
    # _apply needs to be modified also to handle creating instances of UninitializedParameter
    # instead of Parameter

class Linear(nn.Linear, ModuleWithUninitialized):
    def __init__(self, in_features, out_features, bias=True):
        ModuleWithUninitialized.__init__(self)
        self.in_features = in_features
        self.out_features = out_features
        if bias:
            self.bias = Parameter(torch.Tensor(out_features))
        else:
            self.register_parameter('bias', None)
            
        if in_features is INFER:
            self.weight = UninitializedParameter(self._resolve_inferred_input)
        else:
            self.weight = Parameter(torch.Tensor(out_features, in_features))
            self.reset_parameters()
            
    @staticmethod
    def _resolve_inferred_input(self, inputs):
        (input,) = inputs
        self.in_features = input.size()[-1]
        return [self.out_features, self.in_features]
```

I think this will actually have the potential to work in a lot of the cases that it wouldn't have previously. For instance, torch.save should work properly even before forward, which means jit could in theory be made to work also (this might be really valuable down the line for being able to eg. TorchScript torch.hub components that can infer dimensions).

If the logic in ModuleWithUninitialized were placed in nn.Module directly, it wouldn't exactly meet your requirements that performance of non-deferred initialization would be unaffected, because the first pass would have a hook that does a single scan of the local module parameters. This could be mitigated by eg. only adding the hook if register_parameter is called with an UninitializedParameter.

Finally, we need to flesh out what all of the error cases are for accidentally using an uninitialized module (for instance, tracking all of the possible entry points for init-ing parameters) and adding error messages to them.

**Comment by soumith** [comment]
I might not have made my point as clearly last time. I am not excited to see meta-programming or for that matter the use of hooks.

The whole process has to be **explicit**.

I am thinking of something like this:

```python
model = Model(...)

# Should error out in `model.parameters()` if you execute this line, as it encounters UninitializedParameter
optimizer = optim.SGD(model.parameters(), lr = 0.01)  # No error

# finalize by sending an example input
inp = torch.randn(10, 20)
output = model(inp)

# No error
optimizer = optim.SGD(model.parameters(), lr = 0.01)  
```

**Comment by dzhulgakov** [comment]
Second @soumith's point about making it explicit.

One way to prototype something non-intrusively is to create a separate class (or wrapper) for shape inference. Then we can use explicit call to swap out modules for the real ones. E.g.:

```
# constructing model (some code)
model = Sequential([Conv(...), LazyLinear(out_features=10), Relu])

# explicit run of forward (or maybe can be put into init_lazy_modules)
model(torch.randn(...))

# swap up lazy modules in a tree, e.g. replace LazyLinear -> Linear
# errors out if some submodules in the tree didn't see sample inputs (might happen with conditional execution)
model = init_lazy_modules(model)
```

My assumption is that there are likely not that many layers that need lazy evaluation: Linear, RNNs, maybe some attention. So even maintaining `torch.nn.lazy` might not be that big of a deal. Having explicit `init_lazy_modules` call is a bit annoying but makes intent very explicit.

**Comment by vadimkantorov** [comment]
One culprit could also be delayed parameter initialization, and in the worst case - initialization split into eager and delayed

**Comment by yf225** [comment]
Interested to know the result of this discussion, because DPER3 currently has lazy-init (i.e. infer parameter dims and create parameters right before first `forward`), and we are considering adding eager-init features..

**Comment by bethebunny** [comment]
Pytext has been using this implementation:
https://github.com/facebookresearch/pytext/blob/master/pytext/utils/lazy.py

Although I'm not longer actively contributing to the project and I don't know whether they're using it heavily, I have used this implementation for a number of side projects, including sequence classification and some GAN implementations, and they're quite ergonomic.

Basically, `Lazy.partial` operates like `functools.partial`, where it allows you to create a partial function application of an existing module, and pass in some number of positional or keyword arguments. These arguments should be callables which have the same signature of the forward function, and output a value which will be passed in for the actual function application.

`Infer.dimension` is a helper function which is basically `def Infer.dimension(dim): return lambda tensor: tensor.shape[dim]`, which is suitable for most real applications of Lazy modules.

For instance, here's how you would lazy-fy Linear, Conv1d, LayerNorm:
```
from torch import nn
from lazy import Lazy, Infer

LazyLinear = Lazy.partial(nn.Linear, Infer.dimension(-1))
LazyConv1d = Lazy.partial(nn.Conv1d, Infer.dimension(-2))
LazyLayerNorm = Lazy.partial(nn.LayerNorm, Infer.dimension(-1))
```

and you can use LazyLinear eg:
```
# linear will now be a 15 -> 2 dimensional linear layer
linear = lazy.init_lazy_modules(LazyLinear(2), torch.rand(8, 15))
```

This gives you tools to easily lazy-fy more complex modules, for instance
```
class MLP(nn.Module):
  def __init__(self, input_dim, output_dim, dropout=0.2, hidden_dims=(), activation=nn.ReLU, norm=nn.BatchNorm1d):
    dims = [input_dim, *hidden_dims, output_dim]
    self.mlp = nn.Sequential(
        itertools.chain.from_iterable(
            (nn.Linear(in_, out), nn.Dropout(dropout), activation(), norm(out)) for in_, out in zip(dims[:-1], dims[1:])
        )
    )
  def forward(self, t):
    return self.mlp(t)

LazyMLP = Lazy.partial(MLP, Infer.dimension(-1))

my_mlp = lazy.infer_lazy_modules(LazyMLP(2, dropout=0.5, hidden_dims=(128, 64, 32)), torch.rand(6, 20))
```

My main criticism of this implementation is that the first forward pass is now within `init_lazy_modules`, which means when you're developing a new model, most of the failures are in a call stack below this function, so it ends up feeling a bit like a compile step. However, I see it as being slightly superior to the `pytorch-pfn-extras` implementation in a few ways called out in the discussion of this proposal, namely
1. it is quite explicit about how "infer" objects are passed, ie. it doesn't change the existing nn tools at all, but rather makes it easy to create new explicit lazy APIs, and doesn't use `None` or any other symbolic tools to "mock" those APIs
2. it is explicit about how lazy modules are realized. This makes it very clear that a model isn't finished being constructed until after it's passed an example input and explicitly replaced with a non-lazy modules.
3. The realization step explicitly fails if any lazy modules would still exist in the module graph afterwards, ie. if they were somehow gated by the first forward pass.

**Comment by albanD** [comment]
This has been done in https://github.com/pytorch/pytorch/pull/40493
